ostd/src/mm/frame/meta.rs

```diff
@@ -99,9 +100,14 @@ const_assert_eq!(size_of::<MetaSlot>(), META_SLOT_SIZE);
 /// The implemented structure must have a size less than or equal to
 /// [`PAGE_METADATA_MAX_SIZE`] and an alignment less than or equal to
 /// [`PAGE_METADATA_MAX_ALIGN`].
+///
+/// The implementer of the `on_drop` method should ensure that the frame is
+/// safe to be read.
 pub unsafe trait FrameMeta: Any + Send + Sync + Debug + 'static {
     /// Called when the last handle to the page is dropped.
-    fn on_drop(&mut self, _paddr: Paddr) {}
+    fn on_drop(&mut self, reader: VmReader<Infallible>) {
```

Comment:
Good point. The following signature will never be suspicious:

```rust
fn on_drop(&mut self, reader: &mut VmReader<Infallible>)
```

The method would borrow a reader rather than obtaining one.

This method is designed for the LRU recycling. If the LRU manager wants to recycle a frame it would firstly remove the frame from all its owners (VMOs, mappings, etc), and then write the frame's content to disk.

To tell if the frame is the last owning frame by reading the reference count is prone to races: a dirty frame may never be written to disk (for example, two dropping owners see reference counts of 2 simultaneously). So with `on_drop` we would guarantee that the last leave of owner must be observed.

Reply:
The new look of `on_drop` is acceptable to me.

I can't foresee how page swapping should be done. Let's discuss it in the future.